src/nvim/mbyte.c

```diff
@@ -1064,17 +1020,25 @@ int utfc_ptr2len_len(const char_u *p, int size)
  */
 int utf_char2len(int c)
 {
-  if (c < 0x80)
-    return 1;
-  if (c < 0x800)
-    return 2;
-  if (c < 0x10000)
-    return 3;
-  if (c < 0x200000)
-    return 4;
-  if (c < 0x4000000)
-    return 5;
-  return 6;
+  int buf[4] = {0, 0, 0, 0};
+  switch (utf8proc_category(c)) {
+    case UTF8PROC_CATEGORY_LU:
```

Comment:
Looks rather strange: below `utf_isupper` checks whether character is upper-case by comparing category with …_LU. I know for sure that there are non-ASCII uppercase characters, so either `return 1` or `utf_isupper` is incorrect. If utf8proc does not have the corresponding function I would really suggest to keep the current implementation.

BTW, what utf8proc will do if `buf` is NULL?